sopel/irc/capabilities.py

```diff
@@ -0,0 +1,274 @@
+        enabled.
+
+        If the capability is unknown, then its ``available`` attribute will be
+        ``False``, and its ``params`` attribute will be ``None``.
+        """
+        params = self._available.get(name)
+        return CapabilityInfo(
+            name,
+            params,
+            name in self._available,
+            name in self._enabled,
+        )
+
+    @property
+    def available(self) -> Dict[str, Optional[str]]:
+        """Dict of available server capabilities.
+
+        Each key is the name of a capability advertised by the server, and each
+        value is the parameters as advertised (if any) for this capability.
+
+        If a capability is not in this ``dict``, it means the server doesn't
+        advertise it, and it cannot be requested.
+        """
+        return dict(self._available.items())  # return a copy
+
+    @property
+    def enabled(self) -> FrozenSet[str]:
+        """Set of enabled server capabilities.
+
+        Each element is the name of a capability that is enabled on the server.
+        """
+        return frozenset(self._enabled)
+
+    def is_available(self, name: str) -> bool:
+        """Tell if the capability ``name`` is available on the server."""
+        return name in self._available
+
+    def is_enabled(self, name: str) -> bool:
+        """Tell if the capability ``name`` is enabled on the server."""
+        return name in self._enabled
+
+    def handle_ls(self, bot: SopelWrapper, trigger: Trigger) -> bool:
+        """Handle a ``CAP LS`` command.
+
+        This method behaves as a plugin callable with its ``bot`` and
+        ``trigger`` arguments, with the precise goals to handle ``CAP LS``
+        command only, registering available capabilities.
+
+        Then it returns if there is no more ``LS`` command to handle (in case
+        of multi-line ``LS``).
+        """
+        # dev checking
+        assert trigger.event == 'CAP'
+        assert trigger.args[1] == 'LS'
+
+        # extracting capabilities
+        for available_capability in trigger.split():
+            name, *params = available_capability.split('=', maxsplit=1)
+            self._available[name] = params[0] if params else None
+
+        # return if multiline or not
+        multiline = trigger.args[2] == '*'
+        return not multiline
+
+    def handle_ack(
+        self,
+        bot: SopelWrapper,
+        trigger: Trigger,
+    ) -> Tuple[str, ...]:
+        """Handle a ``CAP ACK`` command.
+
+        This method behaves as a plugin callable with its ``bot`` and
+        ``trigger`` arguments, with the precise goals to handle ``CAP ACK``
+        command only, registering enabled or disabled capabilities and tracking
+        acknowledgement.
+
+        Then it returns the list of acknowledged capability requests.
+        """
+        # dev checking
+        assert trigger.event == 'CAP'
+        assert trigger.args[1] == 'ACK'
```

Comment:
Was this supposed to be left in the "production" branch that's ready to merge?

I know some of my past protocol work has shipped `assert` in production code for a few versions, so I can't throw stones at this from my glass house, but just checking.

Reply:
Yes, assert are fine here, I used them to help fellow developers: if they use them in the wrong context, they will trigger `AssertionError` while working on the code. It is very unlikely that someone will use these outside of Sopel's core.